### PR TITLE
refactor: decouple assignment from definition

### LINKED ISSUE #56: Support pre-transferring garbled circuit gates
Right now our GC VM garbles and transfers the gates on demand. Add support for pre-garbling a circuit and transferring the gates to the evaluator ahead of time.

### PR DESCRIPTION
**Pull request by sinui0**
This PR decouples assignment of values from the definition of values, and is the first step needed to support pre-garbling circuits in the offline phase.

# Changes
- Introduced the `PrimitiveType` trait in `mpz-circuits` to help statically distinguish between array values and primitives.
- Renamed `ValueRegistry` and `EncodingRegistry` to `ValueMemory` and `EncodingMemory`
- Updates the `Memory` trait methods, introducing `assign`.
- Removes the ugly and complex `ValueConfig` and `ValueIdConfig` structs
- Moves `ValueId` and `ValueRef` into `mpz-garble` from `mpz-core`
- Introduces an `ArrayRef` type which is a variant of `ValueRef` which enforces some invariants which simplifies logic everywhere.
- Updates `Generator`, `Evaluator` and `DEAP` to accomodate these changes.

### CODE REVIEW

File: garble/mpz-garble/src/protocol/deap/mod.rs [hunk 1]
```diff
@@ -167,13 +166,13 @@ impl DEAP {
             Role::Follower => (id_1, id_0),
         };
 
-        // Setup inputs concurrently.
+        // // Setup inputs concurrently.
```
Comment (th4s) [inline garble/mpz-garble/src/protocol/deap/mod.rs]: Double `//`

**Review by th4s — APPROVED** [review]
Nice :+1:

File: garble/mpz-garble/src/protocol/deap/mod.rs (hunk 1 above)
Comment (sinui0) [reply garble/mpz-garble/src/protocol/deap/mod.rs]:
```suggestion
        // Setup inputs concurrently.
```

**Comment by sinui0** [comment]
@th4s sry I had to make a couple changes after finding issues integrating with tlsn

**Review by th4s — APPROVED** [review]
:+1:

File: garble/mpz-garble/src/generator/mod.rs [hunk 2]
```diff
@@ -100,41 +102,27 @@ impl Generator {
         Ok(())
     }
 
-    /// Setup input values by transferring the encodings to the evaluator
-    /// either directly or via oblivious transfer.
+    /// Transfer encodings for the provided assigned values.
```
Comment (themighty1) [inline garble/mpz-garble/src/generator/mod.rs]:
```suggestion
    /// Transfer active encodings for the provided assigned values.
```

File: garble/mpz-garble/src/memory.rs [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+/// A memory for storing values.
+#[derive(Default)]
+pub struct ValueMemory {
+    /// IDs for each reference
+    id_to_ref: HashMap<String, ValueRef>,
+    /// References for each ID
+    ref_to_id: HashMap<ValueRef, String>,
+    /// Details for each value
+    details: HashMap<ValueId, ValueDetails>,
+    /// Values that have been assigned
```
Comment (themighty1) [inline garble/mpz-garble/src/memory.rs]:
```suggestion
    /// Values that have been assigned and blind values
```

File: garble/mpz-garble/src/memory.rs [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+    /// Get the encoding for a value id if it exists.
+    pub(crate) fn get_encoding_by_id(&self, id: &ValueId) -> Option<EncodedValue<T>> {
+        self.encodings.get(&id.to_u64().into()).cloned()
+    }
+
+    /// Get the encoding for a value if it exists.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the value is an array and if the type of its elements are not consistent.
```
Comment (themighty1) [inline garble/mpz-garble/src/memory.rs]:
```suggestion
    /// Panics if the value is an array and if the type of its elements is not consistent.
```

**Review by themighty1 — APPROVED** [review]
gw

**Pull request merged**